### ISSUE TITLE
 Annotate that params must be of the same type 

### ISSUE THREAD

**Issue by bcremer** [vimeo/psalm#3264]
It would be useful to annotate that two or more params must be of the same type:

```php
/**
 * @psalm-template TCompareType
 * @psalm-param TCompareType $b
 * @psalm-param TCompareType $a
 */
function takesSameType($a, $b): void
{
    assert(gettype($a) === gettype($b));
}

takesSameType('foo', 'bar'); // invalid
takesSameType('foo', 5); // valid
```

This is also useful for the php inbuilt functions. It's already annotatated in the stub file but it's not validated.

https://github.com/vimeo/psalm/blob/686c2d780dfac97b9c4f79b4d6745b5aa7968ae2/src/Psalm/Internal/Stubs/CoreGenericFunctions.phpstub#L223

```php
/**
 * @psalm-template T
 *
 * @param T[] $arr
 * @param callable(T,T):int $callback
 * @param-out list<T> $arr
 */
function usort(array &$arr, callable $callback): bool
{
}
```

```php
$list = [5,6,7];
usort(
    $list, 
    fn(int $a, string $b): int => (int) ($a > $b) // does not complain about the invalid parameter declartion
);
```

https://psalm.dev/r/3dab304d70

This might be related to: https://github.com/vimeo/psalm/issues/742#

**Comment by muglug** [comment]
Psalm will accurately report when both usort closure params are wrong:

```php
$list = [5,6,7];
usort($list, fn(string $a, string $b): int => (int) ($a > $b));
// ERROR: InvalidScalarArgument - 4:1 - Type string should be a subtype of int(5)|int(6)|int(7)
```

The problem with the general case is that Psalm (and Hack, and other PHP static analysis tools) has a different opinion about how templates should be inferred.

For example the function

```
/**
 * @template T
 * @param T $a
 * @param T $b
 * @return T
 */
function pickOne($a, $b) { return rand(0, 1) ? $a : $b; }
```

will union the inputs to get the output type, so `pickOne("hello", 5);` has the type `string|int`.

**Comment by muglug** [comment]
I did fix the failure to detect the `usort` clash
